### DISCUSSION TITLE
Macro/Shortcut to move Z up first then X move

### DISCUSSION THREAD

**Discussion by humneraudio** [vlachoudis/bCNC#1672] [Q&A]
At the moment I have a shortcut to do the following.

G53 Z-2

Its basically my "safe Z" shortcut.

I want to make another shortcut that does this

G53 Z-2 X-410

However, when I trigger this, both Z and X axis move at the same time. What I want is the Z axis to move first and then the X axis move second.

How do I make a shortcut so that the Z axis moves first and then X axis moves after? This shortcut will be my "tool change" shortcut.

**Comment by MARIOBASZ** [comment]
G53 G0 Z-2
G53 G0 X-410

**Reply by MARIOBASZ** [reply]
there is a macro for tool change implemented in bCNC

**Comment by humneraudio** [comment]
How do I add both those commands in as a shortcut though? I tried adding it in like this

G53 G0 Z-2 G53 G0 X-410

But that doesn't work. Throws error 21 “More than one g-code command from same modal group found in block.”

How do I separate both moves?

**Reply by JipeB** [reply]
G53 is a modal function that should only be called once. It remains active until revoked.
A good principle when writing macros is to separate the actions: one action per line to facilitate reading and debugging.

G53 (function call)
G0 Z-2 (Rapid move to Z-2)
X... Y... (Rapid movement to X and Y coordinates)

No need to specify G0 for the X Y movement, the G0 function is also modal; it is revoked by G1, or G2, or G3 (work movements with the speed defined by parameter F...)

**Comment by humneraudio** [comment]
Thanks, however I'm still trying to work out how I make this macro? In the shortcut section I can only add a single gcode line, how do I add more than one action to a shortcut?

**Reply by Harvie** [reply]
![image](https://user-images.githubusercontent.com/276504/152759408-d6eceb1c-7414-412c-a729-1fff0569e102.png)

**Reply by Harvie** [reply]
seems to work for me

**Comment by humneraudio** [comment]
OK, I see, I found this now, so that's under the Control section. If I add my macro here, how do I trigger it with a keyboard or shortcut?

**Reply by JipeB** [reply]
It is explained here. [https://github.com/vlachoudis/bCNC/wiki/Customize-Shortcuts](url)

**Reply by humneraudio** [reply]
It doesn't explain how to trigger a macro with a shortcut though. 

After I create the macro as per Harvie's screenshot, how do I trigger that with a keyboard command?

**Reply by JipeB** [reply]
I don't use keyboard shortcuts but I saw that several routines were implemented in the bCNC.py file (ORI*GIN x y z). I may be wrong but I think the solution would be to take inspiration from a module similar to what you want to do and modify it. The problem is the coding if you don't master the Python language.

Unless I missed something, there is no ability in the current version to switch from a custom macro to a hotkey without coding.

**Reply by humneraudio** [reply]
hmm, yeah I see, I had a look through the code. Adding in an additional command like ORIGIN could work, then I can trigger that via a shortcut.

Further to that. there are also things like this under ControlPage.py

	def go2origin(self, event=None):
		self.sendGCode("G90")
		self.sendGCode("G0Z%d"%(CNC.vars['safe']))
		self.sendGCode("G0X0Y0")
		self.sendGCode("G0Z0")

Then if you take the ORIGIN code as a base...

		# ORI*GIN x y z: move origin to x,y,z by moving all to -x -y -z
		elif rexx.abbrev("ORIGIN",cmd,3):
			try:	dx = -float(line[1])
			except: dx = 0.0
			try:	dy = -float(line[2])
			except: dy = 0.0
			try:	dz = -float(line[3])
			except: dz = 0.0
			self.editor.selectAll()
			self.executeOnSelection("MOVE", False, dx,dy,dz)

to create something like this... ?

		elif rexx.abbrev("CUSTOM",cmd,3):
		self.sendGCode("G90")
		self.sendGCode("G0Z%d"%(CNC.vars['safe']))
		self.sendGCode("G0X-410Y0")
		self.sendGCode("G0Z0")

While I can somewhat read/interpret code, I'm by no means a coder, haha.

Might have to have a play around with it. I run bCNC from Linux so it makes it easy to test code changes at least.

Thanks for the help so far.

**Reply by JipeB** [reply]
Unless one of the coders in charge of bCNC maintenance intervenes, the best solution is probably to experiment with a piece of code. This is what I did to develop the TLO manual tool change procedure.
Indeed, under Linux the update of the code is more flexible. Under Windows, it is necessary for each modification to make a PIP install to test.
Good luck

JP

**Reply by humneraudio** [reply]
So I have found that the Pendant will allow multi line gcode

_/send?gcode=xxx send gcode directly to the machine, multiple lines with %0D_

I tested it and it works. 

I think focusing on a WIFI pendant controller might be a good option. I already have an ESP32 + 20x4 character displaying pulling JSON from bCNC for Status, MPOS and WPOS, it will be simple enough for me to also add buttons to send HTTP requests to bCNC from the same ESP32. It seems almost as responsive as a locally connected keyboard.

**Reply by JipeB** [reply]
I did some tests, entering multi-line macros can be done directly in the .bCNC file which is read each time the software is launched. This is a way to work around the editor which only tolerates one line.
JP

**Reply by humneraudio** [reply]
Can you show me an example? How do I define a new line in the config file?

**Reply by JipeB** [reply]
I made a mistake, it is the bCNC.ini file. There are some native bCNC macros :

[Buttons]
n = 13
name.0 = 0
icon.0 =
tooltip.0 = Aller à 0, 0, 0
command.0 = G90
	G0Z[safe]
	G0X0Y0
	G0Z0
name.1 = Origine XY 
icon.1 = rapid
tooltip.1 = Retour origine XY, sans mouvement en Z
command.1 = G90G0X0Y0
name.2 = Z hors brut
icon.2 = rapid
tooltip.2 = Aller en Z hors brut
command.2 = G90G0Z[safe]
**name.3 = Aller point jauge
icon.3 = rapid
tooltip.3 = Rapide vers le point de jauge
command.3 = G90
	G0Z[safe]
	G53 G0 X[prbx] Y[prby]
	G53 G0 Z[prbz]**
name.4 = Pause
icon.4 = quit
tooltip.4 = Pause grbl (arrêt moteurs sans reset)
command.4 = $SLP
name.5 = Porte
icon.5 = exclamation
tooltip.5 = Ouverture porte
command.5 = SENDHEX 84
name.6 = Jog Stop
icon.6 = rapid
tooltip.6 = Arrêt mouvements manuels (commande $J=)
command.6 = SENDHEX 85
**name.7 = scan
icon.7 = marges
tooltip.7 = Mouvement le long des marges
command.7 = G90
	G0 X[xmin] Y[ymin]
	G0 X[xmax]
	G0 Y[ymax]
	G0 X[xmin]
	G0 Y[ymin]**

Several commands have several lines that can be used for other uses. 
JP
